docs: update NG0506 (#55973)
angular/angular: adev/src/content/reference/errors/NG0506.md

# Application remains unstable

> This warning only appears in the browser during the hydration process when it's enabled on the client but the application remains unstable for an extended period.

Reviewer: ```suggestion
This warning only appears in the browser during the hydration process when it's enabled on the client but the application remains unstable for an extended period of time (over 10 seconds).
```

> The `ApplicationRef.isStable` API uses NgZone to report when an application becomes `stable` and `unstable`. An application is considered stable when there are no pending microtasks or macrotasks.

Reviewer: I think we can rephrase this to drop the mention of NgZone (since it's not used in Zoneless mode):
```suggestion
Typically that happens when there are some pending microtasks or macrotasks on a page.
```

Angular Hydration relies on a signal from `ApplicationRef.isStable` when it becomes stable inside an application:

- during the server-side rendering (SSR) to start the serialization process
- in a browser this signal is used to start the post-hydration cleanup to remove DOM nodes that remained unclaimed

> This warning is displayed when `ApplicationRef.isStable` does not emit `true` within 10 seconds. If this behavior is intentional and your application stabilizes later, you could choose to ignore this warning. However, it is not recommended to maintain long-running tasks in the background that prevent the application from stabilizing.

Reviewer: I'd propose removing the last sentence, since it might be intentional (as mentioned earlier in this paragraph):
```suggestion
This warning is displayed when `ApplicationRef.isStable` does not emit `true` within 10 seconds. If this behavior is intentional and your application stabilizes later, you could choose to ignore this warning.
```

## Applications that use zone.js

> Applications that still use zone.js may encounter various factors contributing to instability. These may include pending HTTP requests, timers, or even `fs.readFile` calls, if it's used during server-side rendering on the Node.js platform, as certain Node.js APIs are also partially patched by zone.js.

Reviewer: I've removed fs and Node-based content, since it's not relevant on the client (where an error happens):
```suggestion
Applications that use zone.js may have various factors contributing to delays in stability. These may include pending HTTP requests, timers (`setInterval`, `setTimeout`) or some logic that continuously invokes `requestAnimationFrame`.
```

### Macrotasks

Macrotasks include functions like `setInterval`, `setTimeout`, `requestAnimationFrame`, etc.

> If one of these functions is called during the initialization phase of the application or in bootstrapped components, it may lead to the application remains unstable.

Reviewer: ```suggestion
If one of these functions is called during the initialization phase of the application or in bootstrapped components, it may delay the moment when the application becomes stable.
```

This warning occurs when hydration is enabled on the client but the NgZone remains unstable for a long period of time.

> The `ApplicationRef.isStable` API uses NgZone to report when an application becomes `stable` and `unstable`. An application is considered stable when there are no pending microtasks or macrotasks.

Reviewer 2: > The `ApplicationRef.isStable` API uses NgZone 
In v18 it uses the internal `PendingTasks` service.

Author: Made updates.

Angular Hydration relies on a signal from Zone.js when it becomes stable inside an application:

- during the server-side rendering (SSR) to start the serialization process
- in a browser this signal is used to start the post-hydration cleanup to remove DOM nodes that remained unclaimed

This warning is displayed when the `ApplicationRef.isStable()` doesn't emit `true` within 10 seconds. If this is intentional and your application becomes stable later, you can ignore this warning.

## Macrotasks

Macrotasks include functions like `setInterval`, `setTimeout`, `requestAnimationFrame`, etc.

> If one of these functions is called during the initialization phase of the application or in bootstrapped components, it may lead to the application remains unstable.

Reviewer 2: Should we make these sentences "ready" for zoneless ?

This warning occurs when hydration is enabled on the client but the NgZone remains unstable for a long period of time.

> The `ApplicationRef.isStable` API uses NgZone to report when an application becomes `stable` and `unstable`. An application is considered stable when there are no pending microtasks or macrotasks. In version 18 and subsequent releases, it uses the internal `PendingTasks` service, with NgZone now contributing to the `PendingTasks`.

Reviewer: I think we can rename the title to "Application remains unstable", i.e. drop "NgZone", since the problem may technically happen in Zoneless mode too. I'd also propose to drop the "In version 18 and subsequent releases, it uses the internal `PendingTasks` service, with NgZone now contributing to the `PendingTasks`." sentence - it's an internal implementation detail (may not have a lot of value to developers debugging a problem). We can mention though that https://angular.dev/api/core/ExperimentalPendingTasks also contribute to the application stability.

  template: ``,
})
class SimpleComponent {
  constructor() {
    setInterval(() => { ... }, 1000)

    // or

    setTimeout(() => { ... }, 10_000)
  }
}
```

> Consider evaluating whether these asynchronous tasks should be invoked during server-side rendering, or if they should only be triggered in the browser. Running timers on the server may result in delayed response times, as the Angular framework waits for the application to stabilize.

Reviewer: This error is thrown in the browser during the hydration process. I'm not sure if mentioning server part of it would be relevant in this guide.

> Invoking those functions outside of the Angular zone on the server is considered a "dirty" hack because it may result in "dangling" timers or other tasks, which should be avoided during server-side rendering. The root view is destroyed after the final HTML is serialized, hence any dangling timers may lead to memory leaks.

Reviewer: Similar to the previous comment - this may not be relevant for this guide (we can find a different location in the docs for this info).

If we end up leaving this paragraph, I'd like to propose an update:
```suggestion
Invoking those functions outside of the Angular zone on the server is not recommended because it may result in "dangling" timers or other tasks, which should be avoided during server-side rendering. The root view is destroyed after the final HTML is serialized, hence any dangling timers may lead to memory leaks.
```

If you need to schedule those timers in the browser, they should also be wrapped with `runOutsideAngular` to allow for hydration cleanup as soon as possible:

```typescript
class SimpleComponent {
  constructor() {
    const platformId = inject(PLATFORM_ID);

    if (isPlatformBrowser(platformId)) {
      inject(NgZone).runOutsideAngular(() => {
        setInterval(() => {}, 1000);
      });
    }
  }
}
```

> When piping streams with `debounceTime` and specifying a timeout to debounce generated events, it's important to note that this practice may contribute to application instability, particularly during server-side rendering. It's advisable to check the current platform and adjust the timeout settings accordingly on the server side:

Reviewer: This also seems to be related to server vs client (see comments above).

```typescript
class SimpleComponent {
  constructor() {
    // Assuming `this.stream$` is either a `Subject` or an
    // `Observable` that emits events...
    const stream$ = this.stream$;
    const platformId = inject(PLATFORM_ID);
    const isBrowser = isPlatformBrowser(platformId);

    stream$.pipe(
      isBrowser ? debounceTime(1000) : debounceTime(0, asapScheduler),
      takeUntilDestroyed()
    ).subscribe(() => {
      // ...
    });
  }
}
```

In the example above, we used `asapScheduler` when the platform is the server. This approach can significantly reduce the duration of application instability.

### Using `afterNextRender`

> You may also resort to using the [`afterNextRender`](api/core/afterNextRender) function, which could provide a cleaner method for executing browser-only code outside of the Angular zone. `afterNextRender` callbacks are automatically invoked outside of the Angular zone and are executed only once, whereas `afterRender` will invoke the provided callback after each render:

Reviewer: This behavior has changed recently, the callbacks are no longer invoke outside of NgZone explicitly: https://github.com/angular/angular/commit/7e89753eeff24f52d39fef92600293bf1700cd1b

```ts
class SimpleComponent {
  constructor() {
    afterNextRender(() => {
      setInterval(() => {}, 1000);
    });
  }
}
```

### Third-party libraries

> Third-party libraries can also be a root cause of "fire and forget" asynchronous tasks. Some of these libraries may schedule tasks during their initialization logic, which can prevent Angular from stabilizing properly. For instance, consider LogRocket, a session replay platform, which may set up a timer using `setTimeout` to send network information. The application will remain unstable until the timer fires. It's important to consider wrapping third-party library code with `runOutsideAngular` as well or run the initialization within the `afterNextRender`:

Reviewer: I'd propose to avoid referencing a specific library, just mention that some third-party libraries may have patterns that would delay application stability. The recommendation might be the same: try running relevant library code outside of zone.

```typescript
import LogRocket from 'logrocket';

class SimpleComponent {
  constructor() {
    const platformId = inject(PLATFORM_ID);

    if (isPlatformBrowser(platformId)) {
      inject(NgZone).runOutsideAngular(() => {
        LogRocket.init(...);
      });
    }

    // Or

    afterNextRender(() => {
      LogRocket.init(...);
    });
  }
}
```

### Running code after the hydration is done

> You may also delay the code that sets up asynchronous tasks until the hydration process is complete to prevent them from delaying the actual hydration process:

Reviewer: That's a good point 👍 I'd propose rephrasing it to say that you can run a code once an application becomes stable (vs "when hydration is done")

```typescript

> import LogRocket from 'logrocket';

Reviewer: Similar to one of the previous comments, let's avoid using a specific library.

class SimpleComponent {
  constructor() {
    const platformId = inject(PLATFORM_ID);

    if (isPlatformBrowser(platformId)) {
      const applicationRef = inject(ApplicationRef);

      applicationRef.isStable.pipe( first((isStable) => isStable) ).subscribe(() => {
        // Note that we don't need to use `runOutsideAngular` because `isStable`
        // emits events outside of the Angular zone when it's truthy (falsy values
        // are emitted inside the Angular zone).
        LogRocket.init(...);
      });
    }
  }
}
```

## Zoneless applications

> Zoneless applications rely solely on the internal `PendingTasks` service. These pending tasks are explicitly added and removed by the Angular framework. They are used in various scenarios, such as within `effect`s, `@defer` blocks for remote resource loading, during navigation by the router, and by the HTTP module, which marks the application as unstable during HTTP requests.

Reviewer: ```suggestion
[Zoneless applications](./guide/experimental/zoneless) rely solely on the internal `PendingTasks` service. These pending tasks are explicitly added and removed by the Angular framework. They are used in various scenarios, such as within `effect`s, `@defer` blocks for remote resource loading, during navigation by the router, and by the HTTP module, which marks the application as unstable during HTTP requests.
```

> In zoneless scenarios, it's more intrinsic to identify the cause of application instability. This might be an `effect` running in an infinite loop (potentially because signals used in effect functions keep changing) or a pending HTTP request.

Reviewer: ```suggestion
In zoneless scenarios, stability might be delayed by an application code inside of an `effect` running in an infinite loop (potentially because signals used in effect functions keep changing) or a pending HTTP request.
```

> Developers may also explicitly contribute to indicating the application's stability by using the experimental `PendingTasks` service.

Reviewer: It'd be great to have a link to API docs for the `ExperimentalPendingTasks` symbol here.

> In conclusion, it's crucial to monitor all asynchronous tasks scheduled within the application and determine whether they are necessary. Additionally, verifying whether third-party libraries use any asynchronous APIs internally is important, as they may contribute to application instability.

Reviewer: I'd propose dropping this paragraph, this would mimic the table of contents on a page and info from the first paragraphs.

  template: ``,
})
class SimpleComponent {
  constructor() {
    setInterval(() => { ... }, 1000)

    // or

    setTimeout(() => { ... }, 10_000)
  }
}
```

> Consider evaluating whether these asynchronous tasks should be invoked during the bootstrap process. Running timers during bootstrap may result in delayed hydration cleanup times, as the Angular framework waits for the application to stabilize.

Reviewer: I'd propose removing this sentence, it looks like it repeats the content mentioned earlier.

> If you need to schedule those timers in the browser, they should also be wrapped with `runOutsideAngular` to allow for hydration cleanup as soon as possible:

Reviewer: We should also rephrase or revert this sentence.

```typescript
class SimpleComponent {
  constructor() {
    const platformId = inject(PLATFORM_ID);

>     if (isPlatformBrowser(platformId)) {

Reviewer: Since the error happens on the client only, there is no need to check for `isPlatformBrowser` here. I'd propose keeping this example simple (similar to the current code).

      const ngZone = inject(NgZone);

      ngZone.runOutsideAngular(() => {
        setInterval(() => {}, 1000);
      });
    }
  }
}
```

### Third-party libraries

> Third-party libraries can also be a root cause of "fire and forget" asynchronous tasks. Some third-party libraries may have patterns that could delay application stability. The recommendation remains the same: try executing relevant library code outside of the zone.

Reviewer: ```suggestion
Some third-party libraries can also produce long-running asynchronous tasks, which may delay application stability. The recommendation is to invoke relevant library code outside of the zone as described above.
```

### Running code after an application becomes stable

You can run a code that sets up asynchronous tasks once an application becomes stable:

```typescript
class SimpleComponent {
  constructor() {
    const platformId = inject(PLATFORM_ID);

>     if (isPlatformBrowser(platformId)) {

Reviewer: Do we need `isPlatformBrowser` in this example?

      const applicationRef = inject(ApplicationRef);

      applicationRef.isStable.pipe( first((isStable) => isStable) ).subscribe(() => {
        // Note that we don't need to use `runOutsideAngular` because `isStable`
        // emits events outside of the Angular zone when it's truthy (falsy values
        // are emitted inside the Angular zone).
        setTimeout(() => { ... });
      });
    }
  }
}
```

## Zoneless applications

> [Zoneless applications](/guide/experimental/zoneless) rely solely on the internal `PendingTasks` service. These pending tasks are explicitly added and removed by the Angular framework. They are used in various scenarios, such as within `effect`s, `@defer` blocks for remote resource loading, during navigation by the router, and by the HTTP module, which marks the application as unstable during HTTP requests.

Reviewer: This is an implementation detail, I'd propose removing this paragraph.

In zoneless scenarios, stability might be delayed by an application code inside of an `effect` running in an infinite loop (potentially because signals used in effect functions keep changing) or a pending HTTP request.

> Developers may also explicitly contribute to indicating the application's stability by using the experimental [`PendingTasks`](/api/core/ExperimentalPendingTasks) service.

Reviewer: ```suggestion
Developers may also explicitly contribute to indicating the application's stability by using the experimental [`PendingTasks`](/api/core/ExperimentalPendingTasks) service. If you use the mentioned APIs in your application, make sure you invoke a function to mark the task as completed.
```
